File: monitor/internal/call.go
package internal

import (
	"crypto/tls"
	"net/http"
	"sync"
	"time"
)

const TIMEOUT = 3 * time.Second

func CallAll(sites []ConfigSite) []*ResultSite {
	var wg sync.WaitGroup
	cR := make(chan *ResultSite)
	for _, site := range sites {
		wg.Add(1)
		go callAsync(wg, cR, site)
	}

	results := []*ResultSite{}
	for range sites {
		result := <-cR
		results = append(results, result)
	}

	return results
}

func callAsync(wg sync.WaitGroup, c chan *ResultSite, site ConfigSite) {
	defer wg.Done()
	res := call(site)

	c <- res
}

func call(site ConfigSite) *ResultSite {
	status := S_DOWN
	method := http.MethodGet
	if isMethod(site.Method) {
		method = site.Method

	}
	request, _ := http.NewRequest(method, site.Url, nil)

	h := &http.Client{
		Timeout: TIMEOUT,
	}
	if site.Insecure {
		h.Transport = &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
		}
	}
	if site.MaxResponseTime > 0 {
		h.Timeout = time.Duration(site.MaxResponseTime)
	}

	timer := time.Now()
	var resp *http.Response
	if request != nil {
		resp, _ = h.Do(request)
	}
	elapsed := time.Since(timer)

	httpCode := 0

	if resp != nil {
		httpCode = resp.StatusCode

		if ok := resp.StatusCode >= 200 && resp.StatusCode < 300; ok {
			status = S_UP
		}
	}

	result := ResultSite{
		Name:         site.Name,
		Status:       status,
		ResponseTime: int(elapsed.Milliseconds()),
		HttpCode:     httpCode,
		Timestamp:    getTimestampNow(),
	}

	return &result
}

func getTimestampNow() string {
	return time.Now().Format(time.RFC3339)
}

func isMethod(m string) bool {
	switch m {
	case http.MethodGet, http.MethodHead, http.MethodPost, http.MethodPut, http.MethodPatch, http.MethodDelete, http.MethodConnect, http.MethodOptions, http.MethodTrace:
		return true
	}
	return false
}

File: monitor/src/call.ts
import got, { OptionsOfTextResponseBody, Method } from "got"
import { ConfigSite, CSVRecord, S_DOWN, S_UP } from "./types"

// //@ts-ignore
// process.env["NODE_TLS_REJECT_UNAUTHORIZED"] = 0

export async function callAll(
	sites: ConfigSite[]
): Promise<Record<string, CSVRecord>> {
	const ws: Promise<CSVRecord>[] = []
	for (let i = 0; i < sites.length; i++) {
		const site = sites[i]

		ws.push(call(site))
	}

	const callRes = await Promise.all(ws)

	let resultDict: Record<string, CSVRecord> = {}
	for (let i = 0; i < sites.length; i++) {
		const site = sites[i]

		resultDict[site.name] = callRes[i]
	}

	return resultDict
}

export async function call(site: ConfigSite): Promise<CSVRecord> {
	let status = S_DOWN
	let method = "get"
	if (site.method) {
		method = site.method
	}
	let timeout = 3000
	if (site.maxResponseTime) {
		timeout = site.maxResponseTime
	}

	const controller = new AbortController()
	const requestConfig: OptionsOfTextResponseBody = {
		method: method as Method,
		timeout,
	}
	if (site.insecure) {
		requestConfig.https = {
			rejectUnauthorized: false,
		}
	}

	const startTime = new Date()
	const response = await got(site.url, requestConfig).catch((err: Error) => err)
	const endTime = new Date()

	const responseTime = endTime.valueOf() - startTime.valueOf()
	if (response instanceof Error) {
		let httpCode = 0
		if (response instanceof got.HTTPError) {
			httpCode = response.response.statusCode
		}
		return {
			status: S_DOWN,
			responseTime,
			httpCode,
			timestamp: startTime.toISOString(),
		}
	}

	if (site.expectedStatusCodes) {
		status = site.expectedStatusCodes.includes(response.statusCode)
			? S_UP
			: S_DOWN
	} else {
		status =
			response.statusCode >= 200 && response.statusCode < 300 ? S_UP : S_DOWN
	}

	return {
		status,
		responseTime,
		httpCode: response.statusCode,
		timestamp: startTime.toISOString(),
	}
}
